### PR TITLE
Update channel_shuffle_op and transpose 2d to speed up ShuffleNet

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: Update channel_shuffle_op and transpose 2d to speed up ShuffleNet

Differential Revision: D8889361

### CODE REVIEW

File: caffe2/utils/math_cpu.cc [hunk 1]
```diff
@@ -2568,24 +2620,53 @@ void TransposeCPUImpl(
...
     const int ndim,
     const int* dims,
     const int* axes,
     const float* X,
-    float* Y,
-    CPUContext* /* context */) {
-#ifdef CAFFE2_USE_HPTT
-  if (TryTransposeWithHPTT(ndim, dims, axes, X, Y)) {
+    float* Y) {
+  if (utils::IsIdentityPermutation(ndim, axes)) {
```
Comment (houseroad) [inline caffe2/utils/math_cpu.cc]: Add another check, if X == Y, just return?

Comment (BIT-silence) [reply caffe2/utils/math_cpu.cc]: Currently transpose cannot support in-place. X must not be Y.

Comment (houseroad) [reply caffe2/utils/math_cpu.cc]: I see, make sense.

**Review by houseroad — APPROVED** [review]
The logic LGTM. Please make sure the iOS part is well tested.

File: caffe2/operators/channel_shuffle_op.cc [hunk 2]
```diff
@@ -1,23 +1,219 @@
...
+    float* Y) {
+  const int stride = G * K * HxW;
+  for (int i = 0; i < N; ++i) {
+    if (G < K) {
+      for (int j = 0; j < G; ++j) {
+        vDSP_mmov(X + j * K * HxW, Y + j * HxW, HxW, K, HxW, G * HxW);
+      }
+    } else {
+      for (int j = 0; j < K; ++j) {
+        vDSP_mmov(X + j * HxW, Y + j * G * HxW, HxW, G, k * HxW, HxW);
```
Comment (houseroad) [inline caffe2/operators/channel_shuffle_op.cc]: typo: k ==> K

**Comment by BIT-silence** [comment]
@pytorch retest this please

**Pull request closed without merging**